### DISCUSSION TITLE
How to use single FAMILY PROVIDER for multiple widgets? Using family provider without passing data?

### DISCUSSION THREAD

**Discussion by devnta** [rrousselGit/riverpod#3488] [General]
Sorry for my bad English.
Assume, I have screen `A` include `B`, `C` and `D` widgets.
Widget tree:
A (parent)
   --- B (child)
   --- C (child)
   --- D (child)
And I have a provider (with family):
```
final myAsyncNotifierProvider = AutoDisposeAsyncNotifierProviderFamily<
    MyAsyncNotifier, MyState, String?>(
  () {
    return MyAsyncNotifier();
  },
);
```

I would like initialize `myAsyncNotifier` in screen `A`.
 `B`, `C` and `D` just need using it.
In `A`:
```
final state = ref.watch(myAsyncNotifierProvider("My Data");
```
My problem is how to use myAsyncNotifierProvider without passing data in `B`, `C` and `D` ? Because all of them just using, not need initialize.

**Comment by shanyuliang** [comment]
I have the same question.

I am currently passing the provider from A to B, C and D by constructor parameter.

**Comment by charlescyt** [comment]
An alternative would be storing the family parameter with a provider.
For example, instead of having a family provider.
```dart
@riverpod
Future<Foo> foo(FooRef ref, {required int id}) async {
  /* fetch foo with id */
  return ...
}
```

We could do
```dart
@riverpod
class SelectedFooId extends _$SelectedFooId {
  @override
  int build() => 1;

  void update(int id) => state = id;
}

@riverpod
Future<Foo> foo(FooRef ref) async {
  final selectedFooId = ref.watch(selectedFooIdProvider);
  /* fetch foo with id */
  return ...
}
```

The documentation here explained combining providers with the `Ref` object instead of using family providers.
https://riverpod.dev/docs/essentials/combining_requests

**Comment by bailyzheng** [comment]
This is a common situation. I want to solve it by passing provider as a parameter to child widget. But I found that AsyncNotifierProvider that with some arguments is not based AsyncNotifierProvider . I did not get the correct base class. I set the parameter type to AutoDisposeAsyncNotifierProvider is OK to AsyncNotifierProvider  that without any arguments. But not work when I add some arguments to the provider. The error is

```
The argument type 'AssetListProvider' can't be assigned to the parameter type 'AutoDisposeAsyncNotifierProviderFamily<AutoDisposeFamilyAsyncNotifier<dynamic, dynamic>, dynamic, dynamic>'.
```

@rrousselGit Any suggestions for us?

**Comment by zhxst** [comment]
Can you show some sample code ? like how do you use B, C and D under A?
If they are children of A, you can pass the state.
Make sure your state is immutable, or the state change may not trigger an update.

**Comment by martipello** [comment]
i was also stuck trying to do this until i realised that this isnt how families work, they kind of do the opposite, families allow for multiple instances of the same provider, the parameters you pass are more like identifiers

final appleProvider = ref.read(myProvider('apple'))
final orangeProvider = ref.read(myProvider('orange'))

so calling ref.read(myProvider('apple')) later in the tree will get me the same provider but ref.read(myProvider('strawberry')) would be a new instance, i refactored mine to be a normal change notifier no params and a public init method which i think @rrousselGit argues against, but all my tests pass 😂 and it only interacts with one widget so if you want to reuse it you reuse the widget 🤷

**Comment by devnta** [comment]
As @martipello say:
> i was also stuck trying to do this until i realised that this isnt how families work, they kind of do the opposite, families allow for multiple instances of the same provider, the parameters you pass are more like identifiers

I had to change my mind
